shared/health/src/main/java/io/pravega/shared/health/HealthEndpoint.java

```diff
@@ -0,0 +1,115 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.shared.health;
+
+import java.util.Map;
+
+/**
+ * The {@link HealthEndpoint} interface defines all the types of requests a {@link HealthService} should expect to serve.
+ */
+public interface HealthEndpoint {
+
+    /**
+     * Provides the name of the default {@link HealthContributor} that all other {@link HealthContributor} objects are
+     * reachable from.
+     *
+     * @return The name/id of the default {@link HealthContributor}.
+     */
+    String getDefaultContributorName();
+
+    /**
+     * Provides the {@link Status} of the {@link HealthContributor} with name {@link #getDefaultContributorName()}.
+     *
+     * @return The {@link Status} of the default {@link HealthContributor}.
+     */
+    default Status getStatus() {
+        return getStatus(getDefaultContributorName());
+    }
+
+    /**
+     * Fetches the {@link Status} of the {@link HealthContributor} mapped to by 'id'.
+     *
+     * @param id The id of some {@link HealthContributor} to search for.
+     * @return The current {@link Status} of the {@link HealthContributor}.
+     */
+    Status getStatus(String id);
+
+    /**
+     * Fetches the {@link Status} of the 'root' {@link HealthContributor}.
+     *
+     * @return The current {@link Status} of the 'root' {@link HealthContributor}.
+     */
+    default boolean isReady() {
+        return isReady(getDefaultContributorName());
+    }
+
+    /**
+     * Determine if some {@link HealthContributor} object is currently in a 'ready' state.
+     *
+     * @param id The id of some {@link HealthContributor} to search for.
+     * @return The 'readiness' result.
+     */
+    boolean isReady(String id);
+
+    /**
+     * Determine if the root {@link HealthContributor} object is currently in an 'alive' state.
+     *
+     * @return The 'liveness' result.
+     */
+    default boolean isAlive() {
+        return isAlive(getDefaultContributorName());
+    }
+
+    /**
+     * Determine if some {@link HealthContributor} object is currently in an 'alive' state.
+     * @param id The id of some {@link HealthContributor} to search for.
+     *
+     * @return The 'liveness' result.
+     */
+    boolean isAlive(String id);
+
+    /**
+     * Fetches the details from the root {@link HealthContributor}. This will always return a {@link Map}
+     * object with no entries.
+     *
+     * @return The {@link Map} of details results.
+     */
+    default Map<String, Object> getDetails() {
+        return getDetails(getDefaultContributorName());
+    }
+
+    /**
+     * Fetches the results from the list of {@link java.util.function.Supplier} provided during {@link HealthContributor}
+     * construction.
+     *
+     * @param id The id of some {@link HealthContributor} to search for.
+     * @return The {@link Map} of details results.
+     */
+    Map<String, Object> getDetails(String id);
```

Comment:
What is the ID to be provided here? How does the operator or any client get to know this id?

Reply:
That is the name/identifier of the `HealthContributor` to request. Correct me if I'm wrong but I'd imagine the operator would only care about the top/service level status, so one would just make a call without the `id` specified. If a client wants to understand what are the available components they just have to make this top level request (`getHealth()`), returning all the health information for the service.

Reply:
The operator for doing the _readiness_ and _liveness_ probes needs to know the **overall** readiness/liveness of the whole component it is probing, so yes you are right it needs to know these values only for the top level component.
I see that HealthEndpoint has both methods `boolean isAlive()` and `boolean isAlive(String id)`, so perhaps the expectation is to invoke `isAlive()` _without args_ to understand the liveness for a top-level component (eg Controller) and provide the `id` only in case of sub-components of Controller. Is my understanding correct?

Reply:
> 
> 
> The operator for doing the _readiness_ and _liveness_ probes needs to know the **overall** readiness/liveness of the whole component it is probing, so yes you are right it needs to know these values only for the top level component.
> I see that HealthEndpoint has both methods `boolean isAlive()` and `boolean isAlive(String id)`, so perhaps the expectation is to invoke `isAlive()` _without args_ to understand the liveness for a top-level component (eg Controller) and provide the `id` only in case of sub-components of Controller. Is my understanding correct?

Thats correct. The parameterized calls are expected to be used mostly by some human operator or some higher level script  or something given that the root is the only `HealthContributor` that can reliably be known to exist ahead of time.